### PR TITLE
Feature/worker join

### PR DESCRIPTION
**Pull request by wesleymonte**

### CODE REVIEW

File: api/worker/handlers.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/ufcg-lsd/arrebol-pb/api"
+	"github.com/ufcg-lsd/arrebol-pb/crypto"
+	"io/ioutil"
+	"net/http"
+	"strings"
+)
+
+const SIGNATURE_HEADER string = "SIGNATURE";
+
+type WorkerSpec struct {
```
Comment (raonismaneoto) [inline api/worker/handlers.go:17]: Do we need to make sure the way the fields are written? I think in the worker it is Id, for example...

File: api/worker/handlers.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	builder.WriteString(fmt.Sprintf("%.2f", t.VCPU))
+	builder.WriteString(separator)
+	builder.WriteString(fmt.Sprint(t.RAM))
+	return builder.String()
+}
+
+func (a *WorkerApi) AddWorker(w http.ResponseWriter, r *http.Request) {
+	err := a.verifySignature(r)
+	if err != nil {
+		api.Write(w, http.StatusUnauthorized, api.ErrorResponse{
```
Comment (raonismaneoto) [inline api/worker/handlers.go:47]: we shouldn't reply if the signature is invalid, just return...

File: api/worker/handlers.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+			Message: err.Error(),
+			Status:  http.StatusUnauthorized,
+		})
+		return
+	}
+
+	var workerSpec WorkerSpec
+	_ = json.NewDecoder(r.Body).Decode(&workerSpec)
+	queueId := a.selectQueue(workerSpec)
```
Comment (raonismaneoto) [inline api/worker/handlers.go:56]: I think this function and the verifySignature one shouldn't be hanging on the api.. What do you think?

File: api/worker/handlers.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	token := a.generateToken(queueId, workerSpec)
```
Comment (raonismaneoto) [inline api/worker/handlers.go:57]: Before generate the token, we should check the workerId in the whitelist..

File: api/worker/handlers.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+	body, _ = r.GetBody()
+	if publicKey, err = crypto.GetWorkerPublicKey(workerSpec.ID); err != nil {return}
+	if message, err = ioutil.ReadAll(body); err != nil {return}
+
+	return crypto.Verify(publicKey, message, []byte(signature))
+}
+
+func (a *WorkerApi) selectQueue(workerSpec WorkerSpec) string {
+	return workerSpec.QueueId
```
Comment (raonismaneoto) [inline api/worker/handlers.go:88]: What if it is nil? (That's what will happen most part of the time)

File: api/worker/handlers.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+	var workerSpec WorkerSpec
+	if err = json.NewDecoder(body).Decode(&workerSpec); err != nil {
+		return errors.New("Maybe the body has a wrong shape")
+	}
+
+	var publicKey *rsa.PublicKey
+	var message []byte
+
+	body, _ = r.GetBody()
+	if publicKey, err = crypto.GetWorkerPublicKey(workerSpec.ID); err != nil {return}
```
Comment (raonismaneoto) [inline api/worker/handlers.go:81]: The worker's publick key is in a very specific format, are you sure that this Get will work? In case of any doubt, check the key_utils file in the worker repo

File: api/worker/handlers.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	if message, err = ioutil.ReadAll(body); err != nil {return}
+
+	return crypto.Verify(publicKey, message, []byte(signature))
```
Comment (raonismaneoto) [inline api/worker/handlers.go:84]: there must be a way to specify the algorithm we used to get the hash of the body, maybe you should check key_utils ...

File: api/worker/handlers.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	if err != nil {
+		api.Write(w, http.StatusUnauthorized, api.ErrorResponse{
+			Message: err.Error(),
+			Status:  http.StatusUnauthorized,
+		})
+		return
+	}
+
+	var workerSpec WorkerSpec
+	_ = json.NewDecoder(r.Body).Decode(&workerSpec)
```
Comment (raonismaneoto) [inline api/worker/handlers.go:55]: check this error and every you haven't checked...

File: test/keys/fake_worker [hunk 9]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (raonismaneoto) [inline test/keys/fake_worker:1]: this keys dir should be in .gitignore

File: api/worker/handlers.go (hunk 1 above)
Comment (wesleymonte) [reply api/worker/handlers.go]: Fixed.

File: api/worker/handlers.go (hunk 2 above)
Comment (wesleymonte) [reply api/worker/handlers.go:34]: What exactly should I return? What is the exitcode?

File: api/worker/handlers.go (hunk 3 above)
Comment (wesleymonte) [reply api/worker/handlers.go]: Yes. Fixed.

File: api/worker/handlers.go (hunk 4 above)
Comment (wesleymonte) [reply api/worker/handlers.go]: Done.

File: api/worker/handlers.go (hunk 5 above)
Comment (wesleymonte) [reply api/worker/handlers.go]: Fixed.

File: api/worker/handlers.go (hunk 6 above)
Comment (wesleymonte) [reply api/worker/handlers.go]: Yes. It's working.

File: api/worker/handlers.go (hunk 8 above)
Comment (wesleymonte) [reply api/worker/handlers.go]: Done.

File: test/keys/fake_worker (hunk 9 above)
Comment (wesleymonte) [reply test/keys/fake_worker:1]: Keys are required to run unit tests.

File: api/worker/api.go [hunk 10]
```diff
@@ -2,17 +2,26 @@ package worker
...
+	"github.com/ufcg-lsd/arrebol-pb/arrebol/worker/manager"
+	"github.com/ufcg-lsd/arrebol-pb/storage"
 	"log"
 	"net/http"
 )
 
 type WorkerApi struct {
 	server  *http.Server
+	manager manager.Manager
+	auth auth.Auth
```
Comment (raonismaneoto) [inline api/worker/api.go:15]: let's call it authenticator, ok? Also, let's try to rename its functions to something more meaningful, for example: What a func called Authenticate should do is to guarantee that the worker is who it is saying it is. The way we verify that is by checking if its signature can be verified with its public key, right? It seems that the Validate func is doing this... We may be mixing concepts here... It is an example, I suggest to rethink all funcs from this module

File: .env.example [hunk 11]
```diff
@@ -4,4 +4,7 @@ DATABASE_PASSWORD=postgres
 DATABASE_PORT=5432
 DATABASE_NAME=postgres
 
-STATIC_WORKER_POOL=5
\ No newline at end of file
+STATIC_WORKER_POOL=5
+WHITE_LIST_PATH=/home/user/go/src/github.com/ufcg-lsd/arrebol-pb/whitelist
+ARREBOL_PRV_KEY_PATH=/home/user/arrebol_key
```
Comment (raonismaneoto) [inline .env.example:9]: Let's use priv, for standardization purposes

File: arrebol/worker/auth/auth.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+)
+
+type Auth interface {
+	Validate(signature []byte, worker *worker.Worker) error
+	Authenticate( worker *worker.Worker) (*token.Token, error)
+	Authorize(token *token.Token) error
+}
+
+type JWTAuth struct {
+	workerKeyReader key.Reader
```
Comment (raonismaneoto) [inline arrebol/worker/auth/auth.go:20]: why a reader and not the key itself?

File: arrebol/worker/whitelist/whitelist.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+type FileWhiteList struct {
+	sourceFile string
+}
+
+func NewFileWhiteList() WhiteList {
+	return &FileWhiteList{sourceFile: os.Getenv(WhiteListPath)}
+}
+
+func (l *FileWhiteList) Contains(workerId string) (bool) {
+	ids, err := l.loadSourceFile()
```
Comment (raonismaneoto) [inline arrebol/worker/whitelist/whitelist.go:26]: you could have a ids filed in the whitelist struct, it will be set in "construction time", so you don't have to load it every time contains is called. In the NewWhiteList you must call loadSourceFIle and assign it to its field, what do you think?

File: arrebol/worker/whitelist/whitelist.go [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+const (
+	WhiteListPath = "WHITE_LIST_PATH"
+)
+
+type WhiteList interface {
+	Contains(id string) (bool)
+}
+
+type FileWhiteList struct {
```
Comment (raonismaneoto) [inline arrebol/worker/whitelist/whitelist.go:17]: Didn't like this name. Let's make things simple, do we really need the WhiteLIst interface? We should call this struct WhiteList, and get rid of the interface

File: api/worker/handlers.go (hunk 2 above)
Comment (raonismaneoto) [reply api/worker/handlers.go:34]: you return nothing, replace the if code block for: `return`. I haven't been clear enough before, sorry.

File: api/worker/api.go [hunk 15]
```diff
@@ -2,17 +2,26 @@ package worker
...
 import (
 	"github.com/gorilla/mux"
+	"github.com/ufcg-lsd/arrebol-pb/arrebol/worker/auth"
+	"github.com/ufcg-lsd/arrebol-pb/arrebol/worker/manager"
+	"github.com/ufcg-lsd/arrebol-pb/storage"
 	"log"
 	"net/http"
 )
 
 type WorkerApi struct {
```
Comment (raonismaneoto) [inline api/worker/api.go:12]: there are many directories, like auth, token, whitelist that give me the feeling they should live in this directory, because they deal with things realated to the worker api. What do you think?

Comment (raonismaneoto) [reply api/worker/api.go:12]: for example, the manager manages the worker's operations, like join...

File: arrebol/worker/key/storage.go [hunk 16]
```diff
@@ -0,0 +1,42 @@
+package key
```
Comment (raonismaneoto) [inline arrebol/worker/key/storage.go:1]: we have this file and the crypto one, but they should be only one, makes sense? And I think it shouldn't be in the worker directory, also it should be as more generic as possible

Comment (raonismaneoto) [reply arrebol/worker/key/storage.go:1]: I don't like this storage name as well..

File: arrebol/worker/whitelist/whitelist.go (hunk 14 above)
Comment (wesleymonte) [reply arrebol/worker/whitelist/whitelist.go:17]: I would not remove the interface. It increases modularization and facilitates possible new white list implementations. What do you think of renaming it to DefaultWhiteList?

File: arrebol/worker/auth/auth.go (hunk 12 above)
Comment (wesleymonte) [reply arrebol/worker/auth/auth.go:20]: It's a reader of all workers' keys

File: .env.example (hunk 11 above)
Comment (wesleymonte) [reply .env.example:9]: Done.

File: api/worker/api.go (hunk 15 above)
Comment (wesleymonte) [reply api/worker/api.go:12]: Well, I'm following the pattern where / api keeps only the definition of routes and the implementation of its handlers. I moved the token and whitelist into the auth package because they are part of authentication domain.

File: arrebol/worker/key/storage.go (hunk 16 above)
Comment (wesleymonte) [reply arrebol/worker/key/storage.go:1]: Crypto.go has generic functions for working with RSA keys. On the other hand, key/storage.go has specific functions to manipulate workers' keys.

Comment (wesleymonte) [reply arrebol/worker/key/storage.go:1]: Didn't you like the file name or the interface name?

File: arrebol/worker/whitelist/whitelist.go (hunk 13 above)
Comment (wesleymonte) [reply arrebol/worker/whitelist/whitelist.go]: Done.

File: arrebol/worker/auth/auth.go (hunk 12 above)
Comment (raonismaneoto) [reply arrebol/worker/auth/auth.go:20]: I got your idea, but we won't need to perform any operation with more than 1 worker key pair. So, I suggest you to remove the storage module, and keep some aux private functions from there in the Authenticator, like the getPubKey (once they're very simple, thanks for crypto.go). I think we're making some kind of over engineering by keeping a very simple module like storage.go. Agreed? Maybe that's not the best explanation, feel free to reach me out in case of any doubt

File: arrebol/worker/whitelist/whitelist.go (hunk 14 above)
Comment (raonismaneoto) [reply arrebol/worker/whitelist/whitelist.go]: I think we won't have another impl of the whiteList. Besides, if we need another implementation we add the interface. I think that my first suggestion is better for now..

File: arrebol/worker/key/storage.go (hunk 16 above)
Comment (raonismaneoto) [reply arrebol/worker/key/storage.go:1]: I didn't like "storage"...

File: api/worker/handlers.go [hunk 17]
```diff
@@ -30,7 +30,7 @@ func (a *WorkerApi) AddWorker(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	if err = a.auth.Validate([]byte(signature), worker); err != nil {
+	if err = a.auth.Authenticate([]byte(signature), worker); err != nil {
```
Comment (raonismaneoto) [inline api/worker/handlers.go:33]: wouldn't we return a partially formed token from Authenticate?

File: arrebol/worker/whitelist/whitelist.go [hunk 18]
```diff
@@ -11,29 +11,31 @@ const (
...
+	list []string
 }
 
 func NewFileWhiteList() WhiteList {
-	return &FileWhiteList{sourceFile: os.Getenv(WhiteListPath)}
-}
-
-func (l *FileWhiteList) Contains(workerId string) (bool) {
-	ids, err := l.loadSourceFile()
+	list, err := loadSourceFile()
```
Comment (raonismaneoto) [inline arrebol/worker/whitelist/whitelist.go]: perfect.

File: api/worker/api.go (hunk 10 above)
Comment (raonismaneoto) [reply api/worker/api.go]: they're improving, nice :)

File: arrebol/worker/auth/token/jwt.go [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+	WorkerId string
+}
+
+type Claims struct {
+	QueueId string `json:"QueueId"`
+	WorkerId string `json:"WorkerId"`
+	jwt.StandardClaims
+}
+
+func NewJWToken(worker *worker.Worker) (*JWToken, error){
```
Comment (raonismaneoto) [inline arrebol/worker/auth/token/jwt.go:33]: please, add a doc explaining this func

Comment (raonismaneoto) [reply arrebol/worker/auth/token/jwt.go:33]: why you return a token with raw, queueId and workerId, if the raw already contains this information?

File: arrebol/worker/whitelist/whitelist.go (hunk 14 above)
Comment (wesleymonte) [reply arrebol/worker/whitelist/whitelist.go]: Done.

File: arrebol/worker/auth/auth.go (hunk 12 above)
Comment (wesleymonte) [reply arrebol/worker/auth/auth.go:20]: Done.

File: api/worker/handlers.go (hunk 17 above)
Comment (wesleymonte) [reply api/worker/handlers.go]: Done.

File: arrebol/worker/whitelist/whitelist.go (hunk 18 above)
Comment (wesleymonte) [reply arrebol/worker/whitelist/whitelist.go]: Thx.

File: arrebol/worker/key/storage.go (hunk 16 above)
Comment (wesleymonte) [reply arrebol/worker/key/storage.go]: Removed.

File: arrebol/worker/auth/token/jwt.go (hunk 19 above)
Comment (wesleymonte) [reply arrebol/worker/auth/token/jwt.go]: I updated JWToken and these fields have been removed.

File: arrebol/worker/auth/token/jwt.go [hunk 20]
```diff
@@ -40,41 +32,72 @@ func NewJWToken(worker *worker.Worker) (*JWToken, error){
...
+	}
+	if err != nil {
+		panic(err)
+	}
 	return false
 }
 
-func (*JWToken) GetPayload(key string) string {
-	return ""
+func (t JWToken) GetPayload(key string) (interface{}, error) {
```
Comment (raonismaneoto) [inline arrebol/worker/auth/token/jwt.go:77]: This name gives the idea that the whole payload is going to be returned, but only one field of it is returned. Maybe you should rename it to GetPayloadField or something better. Same thing for SetPayload

File: arrebol/worker/auth/token/jwt.go [hunk 21]
```diff
@@ -40,41 +32,72 @@ func NewJWToken(worker *worker.Worker) (*JWToken, error){
...
-func (*JWToken) Expired() bool {
+func (t JWToken) Expired() bool {
+	_, err := Parse(t.String())
+	v, _ := err.(*jwt.ValidationError)
+
+	if v.Errors == jwt.ValidationErrorExpired {
+		return true
+	}
+	if err != nil {
+		panic(err)
```
Comment (raonismaneoto) [inline arrebol/worker/auth/token/jwt.go:72]: we don't want to panic here

File: arrebol/worker/auth/token/jwt.go (hunk 20 above)
Comment (wesleymonte) [reply arrebol/worker/auth/token/jwt.go:77]: Done.

File: arrebol/worker/auth/token/jwt.go (hunk 21 above)
Comment (wesleymonte) [reply arrebol/worker/auth/token/jwt.go:72]: Done.

File: arrebol/worker/auth/token/jwt.go [hunk 22]
```diff
@@ -69,9 +69,6 @@ func (t JWToken) Expired() bool {
 	if v.Errors == jwt.ValidationErrorExpired {
 		return true
 	}
-	if err != nil {
```
Comment (raonismaneoto) [inline arrebol/worker/auth/token/jwt.go]: my bad, we don't want to panic here, but log the error please

File: arrebol/worker/auth/token/token.go [hunk 23]
```diff
@@ -0,0 +1,11 @@
+package token
+
+type Token interface {
```
Comment (thiagomanel) [inline arrebol/worker/auth/token/token.go:3]: do you guys plan more than one token implementation?

File: api/handlers.go [hunk 24]
```diff
@@ -170,16 +170,16 @@ func (a *HttpApi) RetrieveQueue(w http.ResponseWriter, r *http.Request) {
...
-			write(w, http.StatusNotFound, ErrorResponse{
+			Write(w, http.StatusNotFound, ErrorResponse{
 				Message: fmt.Sprintf("Queue with ID %d not found", queueID),
 				Status:  http.StatusNotFound,
 			})
 		} else {
 			pendingTasks := a.storage.RetrieveTasksByState(queue.ID, storage.TaskPending)
 			runningTasks := a.storage.RetrieveTasksByState(queue.ID, storage.TaskRunning)
-			response := responseFromQueue(queue, uint(len(pendingTasks)), uint(len(runningTasks)))
+			response := responseFromQueue(queue, uint(len(pendingTasks)), uint(len(runningTasks)), uint(len(queue.Workers)))
```
Comment (thiagomanel) [inline api/handlers.go:180]: why do we have all these concepts of pending, running tasks already in the code? how is this related to the join feature? is this an old codebase?

File: .env.example [hunk 25]
```diff
@@ -4,4 +4,7 @@ DATABASE_PASSWORD=postgres
 DATABASE_PORT=5432
 DATABASE_NAME=postgres
 
-STATIC_WORKER_POOL=5
\ No newline at end of file
+STATIC_WORKER_POOL=5
```
Comment (thiagomanel) [inline .env.example:7]: what is this STATIC_WORKER_POOL? how this is related to the arrebol new design?

File: .env.example [hunk 26]
```diff
@@ -4,4 +4,7 @@ DATABASE_PASSWORD=postgres
...
+WHITE_LIST_PATH=/home/user/go/src/github.com/ufcg-lsd/arrebol-pb/whitelist
```
Comment (thiagomanel) [inline .env.example:8]: what is this WHITE_LIST? how is this related to the arrebol new design?

File: api/worker/handlers.go [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+func (a *WorkerApi) AddWorker(w http.ResponseWriter, r *http.Request) {
+	var err error
+	signature := r.Header.Get(SignatureHeader)
+
+
+	if string(signature) == "" {
+		api.Write(w, http.StatusBadRequest, api.ErrorResponse{
+			Message: "signature header was not found",
```
Comment (thiagomanel) [inline api/worker/handlers.go:24]: rename `signature header was not found` to `Signature header was not found

File: api/worker/handlers.go [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+	ArrebolWorkerToken string
+}
+
+func (a *WorkerApi) AddWorker(w http.ResponseWriter, r *http.Request) {
+	var err error
+	signature := r.Header.Get(SignatureHeader)
+
+
+	if string(signature) == "" {
+		api.Write(w, http.StatusBadRequest, api.ErrorResponse{
```
Comment (thiagomanel) [inline api/worker/handlers.go:23]: I really hate this nested function calls. It is very painful to debug when we need to read the error stacktraces. Is this some sort of idiomatic construct (if so, fine)?

File: arrebol/worker/auth/auth.go [hunk 29]
```diff
@@ -0,0 +1,65 @@
+package auth
```
Comment (thiagomanel) [inline arrebol/worker/auth/auth.go:1]: why is this part of the PR?

File: arrebol/worker/manager/manager.go [hunk 30]
```diff
@@ -0,0 +1,22 @@
+package manager
+
```
Comment (thiagomanel) [inline arrebol/worker/manager/manager.go:2]: is this the name @raonismaneoto suggested in his design? what is a manager?

File: api/handlers.go (hunk 24 above)
Comment (wesleymonte) [reply api/handlers.go:180]: Yes, it is from the old codebase.

File: .env.example (hunk 25 above)
Comment (wesleymonte) [reply .env.example:7]: It is legacy. I removed.

File: arrebol/worker/auth/token/token.go (hunk 23 above)
Comment (wesleymonte) [reply arrebol/worker/auth/token/token.go:3]: No. So I removed that interface.

File: api/worker/handlers.go (hunk 27 above)
Comment (wesleymonte) [reply api/worker/handlers.go:24]: Done.

File: .env.example (hunk 26 above)
Comment (raonismaneoto) [reply .env.example:8]: I've requested @wesleymonte to add this white_list. Btw, we should rename it to avoid discriminatory terms. The meaning of this has been discussed in one of our telcos, and it stands for the list of the workers' ids generated and allowed by the server. This way, only workers whose admins have access to the server can join the server with a valid ID, which avoid some intruders. When we have the ResourceManager, it will be able to acess these ids, and if someone need to deploy a worker by hand, a valid ID is required.

File: arrebol/worker/manager/manager.go (hunk 30 above)
Comment (wesleymonte) [reply arrebol/worker/manager/manager.go]: I updated it. This entity must be an instance of `WorkerScheduler` from diagram.

File: api/worker/handlers.go [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+type TokenResponse struct {
+	ArrebolWorkerToken string
+}
+
+func (a *WorkerApi) AddWorker(w http.ResponseWriter, r *http.Request) {
+	var err error
+	signature := r.Header.Get(SignatureHeader)
+
+
+	if string(signature) == "" {
```
[the remark this answers is not in the document]
Comment (wesleymonte) [reply api/worker/handlers.go]: Done.

File: arrebol/worker/manager/manager.go [hunk 32]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+type Manager struct {
+	storage *storage.Storage
+}
+
+func NewManager(storage *storage.Storage) *Manager {
+	return &Manager{
+		storage:storage,
+	}}
```
Comment (thiagomanel) [inline arrebol/worker/manager/manager.go:15]: out of curiosity, the formatting of this NewManager func is ok?

File: arrebol/worker/manager/manager.go [hunk 33]
```diff
@@ -0,0 +1,34 @@
...
+func (m *Manager) Join(w worker.Worker) (uint, error) {
+	queueId := m.selectQueue(w)
+	w.QueueID = queueId
+	queue, err := m.storage.RetrieveQueue(queueId)
+	if err != nil {
+		return 0, err
+	}
+	queue.Workers = append(queue.Workers, &w)
+	err = m.storage.SaveQueue(queue)
+	if err != nil {
```
Comment (thiagomanel) [inline arrebol/worker/manager/manager.go:26]: `0` means we have problems? I think we are missing docs to this fuction

File: arrebol/worker/manager/manager.go (hunk 30 above)
Comment (thiagomanel) [reply arrebol/worker/manager/manager.go:2]: i try to avoiding using *Manager to name things ... but, ok

File: arrebol/worker/manager/manager.go (hunk 33 above)
Comment (wesleymonte) [reply arrebol/worker/manager/manager.go:30]: What it means that we have a problem is if the err variable is not nil. 0 is just the zero value for the uint type. I added a doc to this function.

File: arrebol/worker/manager/manager.go (hunk 32 above)
Comment (wesleymonte) [reply arrebol/worker/manager/manager.go]: The syntax is ok but the indent was not. Fixed.

File: arrebol/worker/auth/auth.go [hunk 34]
```diff
@@ -0,0 +1,60 @@
...
+	}
+	publicKey, err := key.GetPublicKey(worker.ID)
+	if err != nil {
+		return token, err
+	}
+	err = crypto.Verify(publicKey, data, signature)
+	if err != nil {
+		return token, err
+	}
+	if contains := auth.allowlist.Contains(worker.ID); contains {
```
Comment (thiagomanel) [inline arrebol/worker/auth/auth.go:39]: i think we should not mix the concepts of authorization (backed by this white-list) and authentication

Comment (wesleymonte) [reply arrebol/worker/auth/auth.go]: Okay. I moved the allow list check to the authorization method
